third_party/amd/lib/TritonAMDGPUToLLVM/TargetInfo.cpp

```diff
@@ -126,10 +126,146 @@ Value TargetInfo::programId(RewriterBase &rewriter, Location loc,
   return LLVM::AMD::llGetPid(loc, rewriter, moduleOp, axis);
 }
 
+static std::optional<Operation *> getReduxOpForDPP(triton::ReduceOp op) {
+  // In order to utilize the full power of DPP modifier, the redux operator
+  // should be a single arthmetic operator.
+  //
+  // Reference: third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/TargetInfo.cpp
+  if (op.getNumOperands() != 1 || op.getNumResults() != 1)
+    return std::nullopt;
+  Block *block = &(*op.getCombineOp().begin());
+  Operation *yield = block->getTerminator();
+  Operation *reduceOp = yield->getOperand(0).getDefiningOp();
+  if (!reduceOp || reduceOp->getNumOperands() != 2 ||
+      reduceOp->getNumResults() != 1)
+    return std::nullopt;
+  if (reduceOp->getOperand(0) != block->getArgument(0) ||
+      reduceOp->getOperand(1) != block->getArgument(1))
+    return std::nullopt;
+
+  if (isa<arith::AddIOp>(reduceOp) || isa<arith::AddFOp>(reduceOp) ||
+      isa<arith::AndIOp>(reduceOp) || isa<arith::MaximumFOp>(reduceOp) ||
+      isa<arith::MaxNumFOp>(reduceOp) || isa<arith::MaxSIOp>(reduceOp) ||
+      isa<arith::MaxUIOp>(reduceOp) || isa<arith::MinimumFOp>(reduceOp) ||
+      isa<arith::MinNumFOp>(reduceOp) || isa<arith::MinSIOp>(reduceOp) ||
+      isa<arith::MinUIOp>(reduceOp) || isa<arith::MulFOp>(reduceOp) ||
+      isa<arith::OrIOp>(reduceOp) || isa<arith::SubFOp>(reduceOp) ||
+      isa<arith::SubIOp>(reduceOp) || isa<arith::XOrIOp>(reduceOp)) {
+    return reduceOp;
+  }
+
+  return std::nullopt;
+}
+
 bool TargetInfo::warpReduce(RewriterBase &rewriter, Location loc,
                             SmallVector<Value> &acc, triton::ReduceOp op,
                             unsigned numLaneToReduce,
                             unsigned interleave) const {
+  if (numLaneToReduce != 64)
+    return false;
+
+  if (auto optionalReduxOp = getReduxOpForDPP(op)) {
```

Comment:
In MLIR in general we prefer early return style because it avoids nested indents. So here something like

```c++
Operation *combinerOp = getReduxOpForDPP(op);
if (!cominberOp) return false;
```